tests/Integration/DataProvider/DataProviderWithInMemoryTest.php

```diff
@@ -68,31 +70,48 @@ public function it_can_create_in_memory_factory_in_data_provider(PersistentObjec
 
         $contact = $factory->create();
 
-        self::assertSame([ProxyGenerator::unwrap($contact)], $this->contactRepository->_all());
+        if (TestKernel::canUseLegacyProxy()) {
+            self::assertSame([ProxyGenerator::unwrap($contact)], $this->contactRepository->_all());
+        } else {
+            self::assertSame([$contact], $this->contactRepository->_all());
+        }
 
         self::assertSame(0, $this->entityManager->getRepository(Contact::class)->count());
     }
 
     public static function provideContactFactory(): iterable
     {
         yield [ContactFactory::new()];
-        yield [ProxyContactFactory::new()];
+
+        if (TestKernel::canUseLegacyProxy()) {
+            yield [ProxyContactFactory::new()];
+        }
     }
 
     #[Test]
     #[DataProvider('provideContact')]
     #[AsInMemoryTest]
+    #[Group('legacy-proxy')]
     public function it_can_create_in_memory_objects_in_data_provider(?Contact $contact = null): void
     {
         self::assertInstanceOf(Contact::class, $contact);
 
-        self::assertSame([ProxyGenerator::unwrap($contact)], $this->contactRepository->_all());
+        if (TestKernel::canUseLegacyProxy()) {
+            self::assertSame([ProxyGenerator::unwrap($contact)], $this->contactRepository->_all());
+        } else {
+            self::assertSame([$contact], $this->contactRepository->_all());
+        }
 
         self::assertSame(0, $this->entityManager->getRepository(Contact::class)->count());
     }
 
     public static function provideContact(): iterable
     {
+        if (!TestKernel::canUseLegacyProxy()) {
+            yield [];
+            return;
+        }
```

Comment:
this is currently mandatory in order to still be able to test proxies in data providers

I think I will start working on Foundry 3 very soon because I can't stand this complexity in the tests anymore